consult.el

```diff
@@ -932,6 +940,81 @@ Depending on the selected item OPEN-BUFFER, OPEN-FILE or OPEN-BOOKMARK will be u
   (interactive)
   (consult--buffer #'switch-to-buffer #'find-file #'bookmark-jump))
 
+;;;###autoload
+(defun consult-kmacro (arg)
+  "Run a chosen keyboard macro.  With prefix ARG, run the macro that many times.
+
+Macros containing mouse clicks can't be displayed properly.  To
+keep things simple, macros with an empty display string (e.g.,
+ones made entirely of mouse clicks) are not shown."
+  (interactive "p")
+  (if (or last-kbd-macro kmacro-ring)
+      (let* ((numbered-kmacros
+              (cl-loop
+               ;; The most recent macro is not on the ring, so it must be
+               ;;  explicitly included.
+               for kmacro in (cons (if (listp last-kbd-macro)
```

Comment:
Delete duplicates from the list?

Reply:
I was unsure about removing duplicates, since they are distinct items in the ring. 

Users might want to take advantage of the position of macros in the ring, such as rotating the macro to make switching between macro take less effort. Additionally, because of counters, macros might only be duplicates temporarily.

Reply:
I am not sure what is better. Removing duplicates could be better since after narrowing in the minibuffer the order of the macros is somehow not relevant anymore, since the context is lost. Regarding the counter, you could exclude that from the duplicate check? After selecting the macro are you somehow rotating the ring? If that is the case I think it does not make sense to remove duplicates and you want to present the ring unsorted. But right now I think you do not, you just pick the selected macro and execute it?

Reply:
What I meant was more that adding a duplicate macro is an direct action, so the existence of duplicates can be purposeful/intentional.

However, I now think that your reasoning about special cases applies to this too. People will ask for it, if they want to keep duplicates.

Reply:
I think we can argue here that the interface at hand is simply not meant to work nicely with duplicates. It is just for selecting a macro. Therefore I guess we are good when we delete.